File: solutions/clojure/largest-series-product/2/src/largest_series_product.clj
(ns largest-series-product)

(defn error-message? [span in-str in-chars]
  (cond (neg? span) "span must not be negative"
    (> span (count in-str) ) "span must be smaller than string length" 
    (> (count in-str) (count in-chars)) "digits input must only contain digits"
    :else nil))

(defn largest-product [span input-str] 
  (let [input-chars (clojure.string/split input-str #"(?=[0-9])")
        error (error-message? span input-str input-chars)]
      (when (some? error) (throw (IllegalArgumentException. error)))
         (let [digits (map #(Integer/parseInt %) input-chars)]
           (->
              (for [idx (range (inc (- (count digits) span)))] 
                (reduce * (take span (drop idx digits) )))
              sort
              last))))

File: solutions/dart/largest-series-product/2/lib/largest_series_product.dart
import 'dart:math';

class LargestSeriesProduct {

 

  int largestProduct(String digitString,  int span){
    if (span<0)
      throw ArgumentError('span must not be negative');
    else if (span>digitString.length)
      throw ArgumentError('span must be smaller than string length'); 
    else if (RegExp(r"\D").hasMatch(digitString))
      throw ArgumentError('digits input must only contain digits');
    else if (span==0)
      return 1;
    
    List<int> digits =
      digitString.split('')
       .map((d) => int.tryParse(d))
       .nonNulls
       .toList();    

    int largest=0;
    for(var i=0; i<digits.length - span+1; i++){
      int product = digits.sublist(i,i+span)
        .reduce((a,b)=>a*b);
      if(product>largest) 
        largest = product;
    }
    return largest;
  }
  
}
